### DISCUSSION TITLE
Span / Memory / ReadOnlySequence in C#

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#64] [General]
# Span / Memory / ReadOnlySequence in C#

There are many different memory types used in modern C# programs. The more common ones are Span<T> and Memory<T>. Occasionally there is also ReadOnlySequence<T>. What do these types do?

https://steven-giesel.com/blogPost/8bb02704-a113-40a1-9414-fe7b8cb3dda5

**Comment by ghadzhigeorgiev** [comment]
Thanks for pointing out the **ReadOnlySequence**. Somehow I've missed that one :)

**Reply by linkdotnet** [reply]
To be honest, I did not come across a useful scenario where I needed `ReadOnlySequence`. Did you @ghadzhigeorgiev ?

**Reply by ghadzhigeorgiev** [reply]
If I knew it existed, I could have used it in a project where data was in a file with a fixed size record structure. Could've used it to create a "filtered data reader" with results returned in a nice sequential access to them :)

**Reply by linkdotnet** [reply]
Ahh yeah that would be a really nice use case to "hide" the details and give the client code a nice an clean API. Love it (even though it doesn't help you now :D)

**Comment by gragra33** [comment]
I have looked at using `ReadOnlySequenceSegment<T>` and what is not mentioned, not that I can remember, is that all `ReadOnlyMemory<T>` segment __must be__ the same size. If not, unsized segments are `ascii(0)` filled bytes to match sizes.

**Reply by linkdotnet** [reply]
I can not find any source referencing that fact - do you have any link?

**Reply by gragra33** [reply]
@linkdotnet Documentation on `ReadOnlySequenceSegment<T>` is very limited. The issue is that it is indirectly implied, not formally documented. How it is indirectly implied is via the [ReadOnlySequenceSegment<T>.RunningIndex Property](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequencesegment-1.runningindex?view=net-8.0). There is no tracking for each `ReadOnlyMemory<T>` segment size. it's assumed each is the same size.

I guess that you could track the size of each segment and manually dice and slice. This would kind of defeat the purpose of the `ReadOnlySequenceSegment<T>`  class.

My usage, at the time, was for parsing streams using a [PipeReader](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipelines.pipereader?view=dotnet-plat-ext-8.0) and buffering identified sections. I wanted to use `ReadOnlySequenceSegment<T>` however [PipeReader.ReadAtLeastAsync(Int32, CancellationToken) Method](https://learn.microsoft.com/en-us/dotnet/api/system.io.pipelines.pipereader.readatleastasync?view=dotnet-plat-ext-8.0#system-io-pipelines-pipereader-readatleastasync(system-int32-system-threading-cancellationtoken)) returned variable chunch sizes, This caused the issue described.
